lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+    if (Triple.isOSWindows() && Triple.isArch64Bit()) {
+      // Map UInt to uintptr_t
+      auto swiftUIntType = getNamedSwiftType(stdlib, "UInt");
+      auto clangUIntPtrType = ctx.getCanonicalType(ctx.getUIntPtrType());
+      Cache.insert({swiftUIntType, clangUIntPtrType});
+
+      // Map Int to intptr_t
+      auto swiftIntType = getNamedSwiftType(stdlib, "Int");
+      auto clangIntPtrType = ctx.getCanonicalType(ctx.getIntPtrType());
+      Cache.insert({swiftIntType, clangIntPtrType});
+    }
+    StdlibTypesAreCached = true;
+  }
+
+  auto it = Cache.find(Type(type));
+  if (it != Cache.end())
+    return it->second;
+
+  it = Cache.find(type->getCanonicalType());
+  if (it != Cache.end()) {
+    Cache.insert({Type(type), it->second});
+    return it->second;
+  }
+
+  return clang::QualType();
+}
+
+clang::QualType ClangTypeConverter::visitTupleType(TupleType *type) {
+  unsigned tupleNumElements = type->getNumElements();
+  if (tupleNumElements == 0)
+    return ClangASTContext.VoidTy;
+
+  Type eltTy = type->getElementType(0);
+  for (unsigned i = 1; i < tupleNumElements; i++) {
+    if (!eltTy->isEqual(type->getElementType(i)))
+      // Only tuples where all element types are equal map to fixed-size
+      // arrays.
+      return clang::QualType();
+  }
+
+  auto clangEltTy = convert(eltTy);
+  if (clangEltTy.isNull())
+    return clang::QualType();
+
+  APInt size(32, tupleNumElements);
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
```

Comment:
> How important do we think reliable rejection is for this API? This implementation is definitely converting some types that don't actually have ObjC equivalents.

I don't know 😐. I realized there is an undocumented assumption here which you're kinda' calling out. The contract is:

> We will compute the right Clang type if the Swift type has a corresponding Clang type

not

> We will compute the right Clang type if and only if the Swift type has a corresponding Clang type

We are relying on type-checker and its friends to reject ill-formed`@convention(c)` stuff.

Another somewhat related thing that worries me a bit is that because of the code split (e.g. `ClangTypeConverter` doesn't handle `SILFunctionType` and the like whereas `GenClangType` does), we won't be able to eventually combine the two converters together. Do these two need to have different contracts?

Reply:
Remember that the `GenClangType` converter literally won't exist in a month when we can rely on the Clang function type being present.

Reply:
As for the general question, I think it would be nice to make as much of this code as possible try to do the right thing (i.e. return null) for types without a corresponding Clang type, and then we can call out the places where that's difficult or expensive.

Reply:
Leaving TODOs here instead of adding stronger checking right now.